### PR TITLE
solution

### PR DESCRIPTION
**Pull request by Andriykooo**
[DEMO LINK](https://Andriykooo.github.io/react_goods-selector/)

### CODE REVIEW

File: src/App.jsx [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+          {goodsFromServer.map(item => (
+            <li
+              className={`app__item ${selectedGood.includes(item)
+                && 'app__item--active'
+              }`}
+            >
+              <button
+                type="button"
+                onClick={() => (
+                  selectedGood.includes(item)
```
Comment (victor-arutiunov) [inline src/App.jsx:68]: Good!

**Review by victor-arutiunov — COMMENTED** [review]
Nice work!

File: src/App.jsx [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+  'Bread',
+  'Fish',
+  'Honey',
+  'Jam',
+  'Garlic',
+];
+
+class App extends React.Component {
+  state = {
+    selectedGood: ['Jam'],
```
Comment (mykola-yarchuk) [inline src/App.jsx:19]:
```suggestion
    selectedGoods: ['Jam'],
```

File: src/App.jsx [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+  }
+
+  constructor() {
+    super();
+    this.goods = [...this.state.selectedGood];
```
Comment (mykola-yarchuk) [inline src/App.jsx:24]: there is no reason to that, just save your goods in the state

File: src/App.jsx [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+  }
+
+  selectItem = (item) => {
+    this.goods.push(item);
+    this.setState({ selectedGood: this.goods });
```
Comment (mykola-yarchuk) [inline src/App.jsx:29]: use state only

File: src/App.jsx [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+  }
+
+  removeItem = (item) => {
+    this.goods.splice(this.goods.indexOf(item), 1);
+    this.setState({ selectedGood: this.goods });
```
Comment (mykola-yarchuk) [inline src/App.jsx:34]: only state

File: src/App.jsx [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+  }
+
+  clearItems = () => {
+    this.goods = [];
+    this.setState({ selectedGood: this.goods });
```
Comment (mykola-yarchuk) [inline src/App.jsx:39]: leave only usage of state

File: src/App.jsx [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+  render() {
+    const { selectedGood } = this.state;
+    const goodsIsSelected = selectedGood.length > 0;
+
+    return (
+      <div className="app">
+        <h1>
+          Selected good: -
+          {goodsIsSelected
+            ? `${selectedGood.join(', ')} is selected`
```
[the remark this answers is not in the document]
Comment (mykola-yarchuk) [reply src/App.jsx:51]: good advice from victor-arutinov

File: src/App.jsx [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+              }`}
+            >
+              <button
+                type="button"
+                onClick={() => (
+                  selectedGood.includes(item)
+                    ? this.removeItem(item)
+                    : this.selectItem(item))}
+              >
+                select
```
Comment (mykola-yarchuk) [inline src/App.jsx:72]: you can not only use different methods but also show different text inside the button

**Review by mykola-yarchuk — CHANGES_REQUESTED** [review]

File: src/App.jsx [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+    const preparedTitleGoods = selectedGoods.map((item, index, array) => {
+      if (array.length > 1) {
+        if (index === array.length - 2) {
+          return `${item} and ${array[index + 1]}`;
+        }
+
+        return item;
+      }
+
+      return item;
```
Comment (mykola-yarchuk) [inline src/App.jsx:51]:
```suggestion
      if (array.length > 1 && index === array.length - 2) {
          return `${item} and ${array[index + 1]}`;
      }

      return item;
```

File: src/App.jsx [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+          Selected good: -
+          {goodsIsSelected
+            ? ` ${preparedTitleGoods.length > 1
+              ? preparedTitleGoods.slice(0, selectedGoods.length - 1).join(', ')
+              : preparedTitleGoods.join(', ')} 
+              ${selectedGoods.length === 1
+              ? 'is'
+              : 'are'} selected`
+            : ' No goods selected'
+          }
```
Comment (mykola-yarchuk) [inline src/App.jsx:66]: you can simplify this, it can be implemented easier.

File: src/App.jsx [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+      return item;
+    });
+
+    return (
+      <div className="app">
+        <h1>
+          Selected good: -
+          {goodsIsSelected
+            ? ` ${preparedTitleGoods.length > 1
+              ? preparedTitleGoods.slice(0, selectedGoods.length - 1).join(', ')
```
Comment (mykola-yarchuk) [inline src/App.jsx:60]: em what you trying to achieve with that `slice`??

**Review by mykola-yarchuk — CHANGES_REQUESTED** [review]
good job, but you can simplify it, a lot.

File: src/App.jsx [hunk 12]
```diff
@@ -0,0 +1,94 @@
...
+            ? ` ${selectedGoods.slice(0, selectedGoods.length - 1).join(', ')}
+                ${selectedGoods.length > 1 ? 'and' : ''}
+                ${selectedGoods[selectedGoods.length - 1]}
+                ${selectedGoods.length === 1 ? 'is' : 'are'} selected`
+            : ' No goods selected'
+          }
+        </h1>
+        {goodsIsSelected
+          && <button type="button" onClick={this.clearItems}>X</button>
+        }
```
Comment (mykola-yarchuk) [inline src/App.jsx:57]:
```suggestion
        {goodsIsSelected && (
          <button type="button" onClick={this.clearItems}>
            X
          </button>
        )}
```

File: src/App.jsx [hunk 13]
```diff
@@ -0,0 +1,94 @@
...
+        <ul className="app__list">
+          {goodsFromServer.map(item => (
+            <li
+              key={item}
+              className={`app__item ${selectedGoods.includes(item)
```
Comment (mykola-yarchuk) [inline src/App.jsx:62]: you can create some variable (`isIncluded` for example) inside callback in `map` method before you return `<li>`, and use this variable in all places where you need it. with this approach, you can use the method includes only once

File: src/App.jsx [hunk 14]
```diff
@@ -0,0 +1,94 @@
...
+
+export default App;
+
+// .map((item, index, array) => {
+//   if (index === array.length - 2) {
+//     return `${item} and ${array[index + 1]}`;
+//   };
+
+//   return `${item}`;
+// });
```
Comment (mykola-yarchuk) [inline src/App.jsx:94]: remove that

File: src/App.jsx [hunk 15]
```diff
@@ -0,0 +1,94 @@
...
+    return (
+      <div className="app">
+        <h1>
+          Selected good: -
+          {goodsIsSelected
+            ? ` ${selectedGoods.slice(0, selectedGoods.length - 1).join(', ')}
+                ${selectedGoods.length > 1 ? 'and' : ''}
+                ${selectedGoods[selectedGoods.length - 1]}
+                ${selectedGoods.length === 1 ? 'is' : 'are'} selected`
+            : ' No goods selected'
```
Comment (mykola-yarchuk) [inline src/App.jsx:52]: try to create some extra variables, this mess really hard to read

**Review by mykola-yarchuk — CHANGES_REQUESTED** [review]

**Review by mykola-yarchuk — APPROVED** [review]

**Pull request closed without merging**